lib/cover-project.nix

```diff
@@ -0,0 +1,115 @@
+{ pkgs, stdenv, lib, haskellLib }:
+
+project:
+
+let
+
+  # Create table rows for an project coverage index page that look something like:
+  #
+  # | Package          | TestSuite |
+  # |------------------+-----------|
+  # | cardano-shell    | test-1    |
+  # |                  | test-2    |
+  # | cardano-launcher | test-1    |
+  # |                  | test-2    |
+  #
+  # The logic is a little complex to ensure that only the first test
+  # is listed alongside the package, the second test is accompanied by
+  # a blank 'Package' entry.
+  packageTableRows = package: with lib;
+    let
+      testsOnly = filterAttrs (n: d: isDerivation d) package.components.tests;
+      testNames = mapAttrsToList (testName: _: testName) testsOnly;
+      firstPosition = 1;
+      positions = range firstPosition (length testNames);
+      testNamesWithPosition = zipLists positions testNames;
+      isFirst = pos: pos == firstPosition;
+    in
+      concatStringsSep "\n" (map (tuple:
+        let
+          pos = tuple.fst;
+          testName = tuple.snd;
+        in ''
+      <tr>
+        <td>
+          ${if isFirst pos then ''
+          <a href="${package.identifier.name}-${package.identifier.version}/hpc_index.html">${package.identifier.name}</href>
+          '' else ""} 
+        </td>
+        <td>
+          <a href="${testName}/hpc_index.html">${testName}</a>
+        </td>
+      </tr>
+    '') testNamesWithPosition);
+
+  projectIndexHtml = pkgs.writeText "index.html" ''
+  <html>
+    <head>
+      <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
+    </head>
+    <body>
+      <table border="1" width="100%">
+        <tbody>
+          <tr>
+            <th>Package</th>
+            <th>TestSuite</th>
+          </tr>
+
+          ${with lib; concatStringsSep "\n" (mapAttrsToList (_ : packageTableRows) project)}
+
+        </tbody>
+      </table>
+    </body>
+  </html>
+  '';
+in
+stdenv.mkDerivation {
```

Comment:
I think this is idiomatically a `runCommand` (it's a derivation with just a single build phase)

Reply:
This *can* matter since stdenv depends on the C compiler and `runCommand` doesn't.